### PR TITLE
Tests for BlockTransactions

### PR DESCRIPTION
**Pull request by kushti**

### CODE REVIEW

File: src/test/scala/org/ergoplatform/utils/generators/ErgoTransactionGenerators.scala [hunk 1]
```diff
@@ -275,13 +275,13 @@ trait ErgoTransactionGenerators extends ErgoGenerators with Generators {
   lazy val invalidBlockTransactionsGen: Gen[BlockTransactions] = for {
     headerId <- modifierIdGen
     txs <- Gen.nonEmptyListOf(invalidErgoTransactionGen)
-  } yield BlockTransactions(headerId, txs.foldLeft(Seq.empty[ErgoTransaction])((acc, tx) =>
+  } yield BlockTransactions(headerId, 1: Byte, txs.foldLeft(Seq.empty[ErgoTransaction])((acc, tx) =>
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/utils/generators/ErgoTransactionGenerators.scala:278]: Magic constant?

File: src/test/scala/org/ergoplatform/sanity/ErgoSanityUTXO.scala [hunk 2]
```diff
@@ -91,7 +91,7 @@ class ErgoSanityUTXO extends ErgoSanity[UTXO_ST] with ErgoTestHelpers {
     val boxHolder = boxesHolderGen.sample.get
     val txs = validTransactionsFromBoxHolder(boxHolder)._1
     val id = modifierIdGen.sample.get
-    BlockTransactions(id, txs)
+    BlockTransactions(id, 1: Byte, txs)
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/sanity/ErgoSanityUTXO.scala:94]: Magic constant?

File: src/test/scala/org/ergoplatform/sanity/ErgoSanityDigest.scala [hunk 3]
```diff
@@ -97,6 +97,6 @@ class ErgoSanityDigest extends ErgoSanity[DIGEST_ST] {
     val boxHolder = boxesHolderGen.sample.get
     val txs = validTransactionsFromBoxHolder(boxHolder)._1
     val id = modifierIdGen.sample.get
-    BlockTransactions(id, txs)
+    BlockTransactions(id, 1: Byte, txs)
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/sanity/ErgoSanityDigest.scala:100]: Magic constant.

File: src/test/scala/org/ergoplatform/nodeView/state/UtxoStateSpecification.scala [hunk 4]
```diff
@@ -75,7 +75,8 @@ class UtxoStateSpecification extends ErgoPropertyTest with ErgoTransactionGenera
         height = height,
         parentId = us.stateContext.lastHeaderOpt.map(_.id).getOrElse(Header.GenesisParentId))
       val adProofs = ADProofs(realHeader.id, adProofBytes)
-      val fb = ErgoFullBlock(realHeader, BlockTransactions(realHeader.id, txs), genExtension(realHeader, us.stateContext), Some(adProofs))
+      val bt = BlockTransactions(realHeader.id, 1: Byte, txs)
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/nodeView/state/UtxoStateSpecification.scala:78]: Magic constant.

File: src/test/scala/org/ergoplatform/nodeView/history/BlockSectionValidationSpecification.scala [hunk 5]
```diff
@@ -24,9 +56,9 @@ class BlockSectionValidationSpecification extends HistoryTestHelpers {
     commonChecks(history, block.extension, block.header)
   }
 
-  private def init() = {
+  private def init(version: Version = Header.InitialVersion) = {
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/nodeView/history/BlockSectionValidationSpecification.scala:59]: I suggest to always pass version explicitly.

File: src/test/scala/org/ergoplatform/nodeView/history/BlockSectionValidationSpecification.scala [hunk 6]
```diff
@@ -4,12 +4,44 @@ import org.ergoplatform.modifiers.BlockSection
...
+  
+  private def changeProofByte(version: Version, outcome: Symbol) = {
+    val (history, block) = init(version)
+    val bt = block.blockTransactions
+    val txBytes = HistoryModifierSerializer.toBytes(bt)
 
-  property("BlockTransactions validation") {
+    val txs = bt.transactions
+    val proof = txs.head.inputs.head.spendingProof.proof
+    proof(0) = if(proof.head < 0) (proof.head + 1).toByte else (proof.head - 1).toByte
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/nodeView/history/BlockSectionValidationSpecification.scala:22]: What this means?

File: src/main/scala/org/ergoplatform/modifiers/history/Header.scala [hunk 7]
```diff
@@ -172,7 +177,7 @@ object Header extends ApiCodecs {
       votes = header.votes.toColl
     )
 
-  val CurrentVersion: Byte = 1
+  val InitialVersion: Byte = 1
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/history/Header.scala:180]: ScalaDoc

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala [hunk 8]
```diff
@@ -20,11 +21,13 @@ import scorex.util.Extensions._
 /**
   * Section of a block which contains transactions.
   *
-  * @param headerId - identifier of a header of a corresponding block
-  * @param txs - transactions of a block
-  * @param sizeOpt - (optional) size of the section (cached to not be calculated again)
+  * @param headerId     - identifier of a header of a corresponding block
+  * @param blockVersion - block version
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:25]: what is block version?

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala [hunk 9]
```diff
@@ -35,10 +38,17 @@ case class BlockTransactions(headerId: ModifierId,
 
   lazy val txIds: Seq[Array[Byte]] = txs.map(_.serializedId)
 
+  //todo: add extra byte to separate from txIds?
+  lazy val witnessIds: Seq[Array[Byte]] = txs.map(_.witnessSerializedId)
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:42]: Please add ScalaDocs for public methods.

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala [hunk 10]
```diff
@@ -98,28 +117,49 @@ object BlockTransactions extends ApiCodecs {
...
 }
 
 object BlockTransactionsSerializer extends ScorexSerializer[BlockTransactions] {
-
-  override def serialize(obj: BlockTransactions, w: Writer): Unit = {
-    w.putBytes(idToBytes(obj.headerId))
-    w.putUInt(obj.txs.size)
-    obj.txs.foreach { tx =>
+  // See a comment in the parse() function
+  val MagicNumber = 1000000
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:128]: Let's call it like `MaxTransactionsInV1Block`

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala [hunk 11]
```diff
@@ -98,28 +117,49 @@ object BlockTransactions extends ApiCodecs {
...
+      */
+    val (blockVersion, txCount) = if (verOrCount > MagicNumber) {
+      ((verOrCount - MagicNumber).toByte, r.getUInt().toIntExact)
+    } else {
+      (1: Byte, verOrCount)
+    }
+
+    val txs = (1 to txCount).map { _ =>
       ErgoTransactionSerializer.parse(r)
     }
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:162]: Please use cfor, it is even more natural here.

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala [hunk 12]
```diff
@@ -98,28 +117,49 @@ object BlockTransactions extends ApiCodecs {
...
+      * than 1,000,000 transactions.
+      *
+      * Then the new serializer puts 1,000,000 + block version (while the old one just puts tx count with no version),
+      * and the reader knows that a new serializer was used if the first unsigned integer read is more than 1,000,000.
+      */
+    val (blockVersion, txCount) = if (verOrCount > MagicNumber) {
+      ((verOrCount - MagicNumber).toByte, r.getUInt().toIntExact)
+    } else {
+      (1: Byte, verOrCount)
+    }
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:158]: Please use vars, avoid tuple allocation

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala (hunk 10 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:128]: done

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala (hunk 11 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:162]: Not related to this PR, please make issue to do this consistently around serializers

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala (hunk 12 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:158]: done

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala (hunk 8 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:25]: protocol version for the bloc, changed desc a bit

File: src/test/scala/org/ergoplatform/nodeView/history/BlockSectionValidationSpecification.scala (hunk 5 above)
Comment (kushti) [reply src/test/scala/org/ergoplatform/nodeView/history/BlockSectionValidationSpecification.scala:59]: for testing code this is okay I guess

File: src/test/scala/org/ergoplatform/nodeView/state/UtxoStateSpecification.scala (hunk 4 above)
Comment (kushti) [reply src/test/scala/org/ergoplatform/nodeView/state/UtxoStateSpecification.scala:79]: for testing code this is okay I guess

File: src/test/scala/org/ergoplatform/sanity/ErgoSanityDigest.scala (hunk 3 above)
Comment (kushti) [reply src/test/scala/org/ergoplatform/sanity/ErgoSanityDigest.scala:101]: for testing code this is okay I guess

File: src/test/scala/org/ergoplatform/utils/generators/ErgoTransactionGenerators.scala (hunk 1 above)
Comment (kushti) [reply src/test/scala/org/ergoplatform/utils/generators/ErgoTransactionGenerators.scala:280]: for testing code this is okay I guess

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala (hunk 9 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala:47]: done

**Pull request merged**